### PR TITLE
curp-validations and users

### PR DESCRIPTION
**Pull request by chiqeen03**
To create a user, first we need to validate their CURP by using the resource `CurpValidation`. This resource will create a `curp_validation` object in the database, which will be later fetched by the `User` resource.

We create the CurpValidation by sending information like name, date of birth, state of birth, country of birth, gender, and if we want we can pass the CURP itself so it is validated as well. All these information will be used to generate and validate the CURP with renapo.

```python
curp_validation_request = CurpValidationRequest(
    names='Pedro',
    first_surname='Páramo',
    date_of_birth='1999-01-01',
    state_of_birth='DF',
    country_of_birth='MX',
    gender='male',
)
curp_valdation = cuenca.CurpValidation.create(curp_validation_request)
```
After the `curp_validation` is created, an Identity is also created in the same process. This Identity will be shared across the multiple platforms, and contains information about the physical person.

For the creation of the `User`, we will use the information related to the account for the platform, and also the CURP in order to match the `User` with the `CurpValidation`. In case the process cannot find a valid CurpValidation object in the database, it will return an instruction to use the `curp_validation` resource.

```python
user_request = UserRequest(
    curp='AATM970329HDFVVN05',
    phone_number='+525559610838',
    email_address='manuel@cuenca.com',
    profession='employee',
    address=dict(
        street='calle 1',
        ext_number='2',
        int_number='3',
        postal_code='09900',
        state=State.DF,
        country='MEX',
    ),
)
user = cuenca.User.create(user_request)
```

Some fields like the terms of service, and all the KYC Files, will be asked later through the onboarding, and will be added to the model through the `User.update` method.

```python
user_id = 'USluZkZ7odQvOwelWLLrFH5w'
changes = dict(
    profession='programmer',
    phone_number='+525555555555',
    govt_id=dict(
        type='ine',
        uri_front='cuenca.com',
        uri_back='cuenca.com',
        is_mx=True,
    ),
)
user = User.update(user_id, **changes)
```

### CODE REVIEW

File: cuenca/resources/users.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+class User(Creatable, Retrievable, Updateable):
+    _resource: ClassVar = 'users'
+
+    id: str
+    identity_uri: str
+    platform_id: str
+    created_at: dt.datetime
+    updated_at: dt.datetime
+    phone_number: PhoneNumber
+    email_address: str
```
Comment (matin) [inline cuenca/resources/users.py:32]: `EmailStr` from pydantic

File: cuenca/resources/base.py [hunk 2]
```diff
@@ -194,3 +195,13 @@ class Wallet(Creatable, Deactivable, Retrievable, Queryable):
     @property
     def wallet_uri(self):
         return f'/{self._resource}/{self.id}'
+
+
+class Event(Retrievable, Queryable):
+    _resource: ClassVar
```
Comment (rogelioLpz) [inline cuenca/resources/base.py:201]: Esta linea no es necesaria. Ya viene en el resource Base

File: cuenca/resources/base.py [hunk 3]
```diff
@@ -194,3 +195,13 @@ class Wallet(Creatable, Deactivable, Retrievable, Queryable):
...
+
+    id: str
+    identity_id: str
+    new_model: dict
+    type: EventType
+    created_at: dt.datetime
```
Comment (rogelioLpz) [inline cuenca/resources/base.py:207]: `created_at`  ya esta en Queryable

File: cuenca/resources/base.py [hunk 4]
```diff
@@ -194,3 +195,13 @@ class Wallet(Creatable, Deactivable, Retrievable, Queryable):
     @property
     def wallet_uri(self):
         return f'/{self._resource}/{self.id}'
+
+
+class Event(Retrievable, Queryable):
```
Comment (rogelioLpz) [inline cuenca/resources/base.py:200]: Y si esto lo defines como `IdentityEvent` y  haces que `UserEvent`extienda de el ?

File: cuenca/resources/curp_validations.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    first_surname: str
+    second_surname: Optional[str]
+    date_of_birth: dt.date
+    country_of_birth: Optional[str]
+    state_of_birth: State
+    gender: Gender
+    nationality: Optional[str]
+    manual_curp: Optional[str]
+    calculated_curp: str
+    validated_curp: Optional[str]
```
Comment (rogelioLpz) [inline cuenca/resources/curp_validations.py:26]: Revisar bien tipo de dato de esta variable

File: cuenca/resources/identities.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    State,
+    TOSAgreement,
+)
+from pydantic.dataclasses import dataclass
+
+from .base import Queryable, Retrievable, Updateable
+
+
+@dataclass
+class Identity(Retrievable, Updateable, Queryable):
```
Comment (rogelioLpz) [inline cuenca/resources/identities.py:18]: porque es `Updateable` ?

File: cuenca/resources/identities.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    _resource: ClassVar = 'identities'
+    _query_params: ClassVar = IdentityQuery
+
+    id: str
+    created_at: dt.datetime
```
Comment (rogelioLpz) [inline cuenca/resources/identities.py:23]: Ya estan en los modelos Base

File: cuenca/resources/identities.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    updated_at: dt.datetime
+    names: str
+    first_surname: str
+    second_surname: Optional[str]
+    curp: Optional[str]
```
Comment (rogelioLpz) [inline cuenca/resources/identities.py:28]: type Curp

File: cuenca/resources/base.py [hunk 9]
```diff
@@ -194,3 +195,13 @@ class Wallet(Creatable, Deactivable, Retrievable, Queryable):
...
+    _resource: ClassVar
+
+    id: str
```
Comment (rogelioLpz) [inline cuenca/resources/base.py:203]: Veo que en **TODOS**  los resources agregas este campo. Ya algunos no son ncesarios si extiendes de las clases
`id`  -> Resource
`created_at` -> Queryable
`updated_at` -> Updatable

File: cuenca/resources/users.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+from pydantic.dataclasses import dataclass
+
+from ..http import Session, session as global_session
+from .base import Creatable, Retrievable, Updateable
+from .identities import Identity
+from .resources import retrieve_uri
+
+
+@dataclass
+class User(Creatable, Retrievable, Updateable):
```
Comment (rogelioLpz) [inline cuenca/resources/users.py:26]: Este no va a ser `Queryable` ?

File: tests/resources/test_users.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    user_id = 'USluZkZ7odQvOwelWLLrFH5w'
+    user = User.retrieve(user_id)
+    assert user_id == user.id
+
+
+@pytest.mark.vcr
+def test_update_user():
+    user_id = 'USluZkZ7odQvOwelWLLrFH5w'
+    changes = dict(
+        profession='programmer',
```
Comment (rogelioLpz) [inline tests/resources/test_users.py:28]: Va a haber un catalogo de `proffessions` ?

File: tests/resources/test_users.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    curp_validation = CurpValidation.create(curp_validation_request)
+    assert curp_validation.renapo_curp_match
+    # creating the user
+    user = User.create(user_request)
+    assert user.id is not None
+    assert user.level == 1
+
+
+@pytest.mark.vcr
+def test_retrieve_user():
```
Comment (rogelioLpz) [inline tests/resources/test_users.py:18]: Igual y borra este. Ya lo incluyes en `test_retrieve_user_identity`

**Review by rogelioLpz — COMMENTED** [review]
Sería bueno poner un test donde se valide el flujo completo.
Por ejemplo este [test](https://github.com/cuenca-mx/cuenca-python/blob/main/tests/resources/test_wallet_transactions.py#L46) que se puso para Ahorros

File: cuenca/resources/identities.py (hunk 6 above)
Comment (chiqeen03) [reply cuenca/resources/identities.py:18]: se pueden cambiar datos de una identity, solo dependería de los permisos de api keys según entendí

Comment (rogelioLpz) [reply cuenca/resources/identities.py]: Se borra Updateable

File: tests/resources/test_users.py (hunk 11 above)
Comment (rogelioLpz) [reply tests/resources/test_users.py:27]: Por ahora dice el @pachCode  que nel

File: cuenca/resources/curp_validations.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    manual_curp: Optional[CurpField]
+    calculated_curp: CurpField
+    validated_curp: Optional[CurpField]
+    renapo_curp_match: Optional[bool]
+    renapo_full_match: Optional[bool]
+
+    @classmethod
+    def create(
+        cls,
+        curp_validation_request: CurpValidationRequest,
```
Comment (matin) [inline cuenca/resources/curp_validations.py:33]: nope. this breaks the pattern used in other places: https://github.com/cuenca-mx/cuenca-python/blob/main/cuenca/resources/transfers.py#L37-L44

**Review by matin — CHANGES_REQUESTED** [review]

File: cuenca/resources/users.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    identity_uri: str
+    platform_id: str
+    level: int
+    phone_number: PhoneNumber
+    email_address: EmailStr
+    profession: str
+    terms_of_service: Optional[TOSAgreement]
+    status: Optional[UserStatus]
+    address: Optional[Address]
+    govt_id: Optional[KYCFile]
```
Comment (matin) [inline cuenca/resources/users.py:40]: shouldn't this be a list of `KYCFile` to support front and back?

File: cuenca/resources/users.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    ) -> 'User':
+        return cast(
+            'User', cls._create(session=session, **user_request.dict())
+        )
+
+    @classmethod
+    def update(
+        cls,
+        user_id: str,
+        phone_number: Optional[str] = None,
```
Comment (rogelioLpz) [inline cuenca/resources/users.py:60]: type PhoneNumber igual en el UserUpdateRequest

File: cuenca/resources/curp_validations.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+from pydantic.dataclasses import dataclass
+
+from ..http import Session, session as global_session
+from .base import Creatable, Retrievable
+
+
+@dataclass
+class CurpValidation(Creatable, Retrievable):
+    _resource: ClassVar = 'curp_validations'
+
```
Comment (rogelioLpz) [inline cuenca/resources/curp_validations.py:15]: Falta `created_at` no lo tienen los Resources Base

File: tests/resources/cassettes/test_create_curp_validations.yaml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+      - application/json
+      User-Agent:
+      - cuenca-python/0.7.15.dev1
+      X-Cuenca-Api-Version:
+      - '2020-03-19'
+    method: POST
+    uri: https://sandbox.cuenca.com/curp_validations
+  response:
+    body:
+      string: '{"id":"CVKTTRnlVtTVaAh7bMfV4P0Q","names":"Manuel","first_surname":"Avalos","second_surname":"Tovar","date_of_birth":"1997-03-29","country_of_birth":"MEX","state_of_birth":"DF","gender":"male","nationality":null,"manual_curp":null,"calculated_curp":"AATM970329HDFVVN05","validated_curp":null,"renapo_curp_match":true,"renapo_full_match":false}'
```
Comment (matin) [inline tests/resources/cassettes/test_create_curp_validations.yaml:27]: @chiqeen03 the `country_of_birth` in the response isn't ISO 3166 Alpha-2

File: tests/resources/cassettes/test_create_identity.yaml [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+      - application/json
+      User-Agent:
+      - cuenca-python/0.7.15.dev1
+      X-Cuenca-Api-Version:
+      - '2020-03-19'
+    method: POST
+    uri: https://sandbox.cuenca.com/curp_validations
+  response:
+    body:
+      string: '{"id":"CVy1LipfRRTc6sF5nAFjtlmg","names":"Manuel","first_surname":"Avalos","second_surname":"Tovar","date_of_birth":"1997-03-29","country_of_birth":"MEX","state_of_birth":"DF","gender":"male","nationality":null,"manual_curp":null,"calculated_curp":"AATM970329HDFVVN05","validated_curp":null,"renapo_curp_match":true,"renapo_full_match":false}'
```
Comment (matin) [inline tests/resources/cassettes/test_create_identity.yaml:27]: why is `nationality: null` and `renapo_full_match: false`?

File: cuenca/resources/curp_validations.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    date_of_birth: dt.date
+    country_of_birth: Optional[str]
+    state_of_birth: State
+    gender: Gender
+    nationality: Optional[str]
+    manual_curp: Optional[CurpField]
+    calculated_curp: CurpField
+    validated_curp: Optional[CurpField]
+    renapo_curp_match: Optional[bool]
+    renapo_full_match: Optional[bool]
```
Comment (matin) [inline cuenca/resources/curp_validations.py:28]: no deben ser `Optional`

File: cuenca/resources/curp_validations.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    @classmethod
+    def create(
+        cls,
+        names: str,
+        first_surname: str,
+        date_of_birth: dt.date,
+        country_of_birth: str,
+        gender: Gender,
+        second_surname: Optional[str] = None,
+        state_of_birth: Optional[State] = None,
```
Comment (matin) [inline cuenca/resources/curp_validations.py:40]: por qué es opcional?

**Review by matin — CHANGES_REQUESTED** [review]
@felipaoo acaba de hacer una prueba de `curp_validations`, y recibí un error

File: cuenca/resources/curp_validations.py [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+@dataclass
+class CurpValidation(Creatable, Retrievable):
+    _resource: ClassVar = 'curp_validations'
+
+    created_at: dt.datetime
+    names: str
+    first_surname: str
+    second_surname: Optional[str]
+    date_of_birth: dt.date
+    country_of_birth: Optional[str]
```
[the remark this answers is not in the document]
Comment (chiqeen03) [reply cuenca/resources/curp_validations.py:21]: error mio

File: cuenca/resources/curp_validations.py (hunk 20 above)
Comment (chiqeen03) [reply cuenca/resources/curp_validations.py:40]: El state solo es requerido cuando el country of birth es MX

File: cuenca/resources/users.py (hunk 14 above)
Comment (pachCode) [reply cuenca/resources/users.py:42]: El KYCFile tiene los atributos `front_uri` y `back_uri` por lo que no es necesario
https://github.com/cuenca-mx/cuenca-validations/blob/49f1eeb6df5ea717a3fd3cef5531a329c95137b7/cuenca_validations/types/identities.py#L47

Comment (pachCode) [reply cuenca/resources/users.py:42]: @matin   ^^

File: requirements.txt [hunk 22]
```diff
@@ -1,3 +1,3 @@
-requests==2.27.1
-cuenca-validations==0.9.16
+requests==2.26.0
```
Comment (matin) [inline requirements.txt:1]: why are you using an old version?

File: setup.py [hunk 23]
```diff
@@ -23,8 +23,8 @@
     package_data=dict(cuenca=['py.typed']),
     python_requires='>=3.6',
     install_requires=[
-        'requests>=2.24,<2.28',
-        'cuenca-validations>=0.9.0,<0.10.0',
+        'requests>=2.24,<2.27',
+        'cuenca-validations>=0.9.18.dev3,<0.9.19',
```
Comment (matin) [inline setup.py:27]: you shouldn't reference dev versions of a library

File: setup.py [hunk 24]
```diff
@@ -23,8 +23,8 @@
     package_data=dict(cuenca=['py.typed']),
     python_requires='>=3.6',
     install_requires=[
-        'requests>=2.24,<2.28',
-        'cuenca-validations>=0.9.0,<0.10.0',
+        'requests>=2.24,<2.27',
```
Comment (matin) [inline setup.py:26]: why are you not allowing the latest version of `requests`?

File: tests/conftest.py [hunk 25]
```diff
@@ -23,3 +27,36 @@ def transfer():
...
+
+
+@pytest.fixture
+def curp_validation_request() -> Dict:
+    curp_validation = dict(
+        names='José',
+        first_surname='López',
+        second_surname='Hernández',
+        date_of_birth=dt.date(1966, 6, 6).isoformat(),
+        state_of_birth=State.DF.value,
```
Comment (matin) [inline tests/conftest.py:39]: remove `.value`

File: tests/conftest.py [hunk 26]
```diff
@@ -23,3 +27,36 @@ def transfer():
...
     )
+
+
+@pytest.fixture
+def curp_validation_request() -> Dict:
+    curp_validation = dict(
+        names='José',
+        first_surname='López',
+        second_surname='Hernández',
+        date_of_birth=dt.date(1966, 6, 6).isoformat(),
```
Comment (matin) [inline tests/conftest.py:38]: shouldn't it convert it to isoformat automatically when you make the request?

**Review by matin — CHANGES_REQUESTED** [review]

File: requirements.txt [hunk 27]
```diff
@@ -1,3 +1,3 @@
-requests==2.27.1
-cuenca-validations==0.9.16
+requests==2.27.0
+cuenca-validations==0.9.18.dev10
```
Comment (matin) [inline requirements.txt:2]: why?

File: setup.py [hunk 28]
```diff
@@ -23,8 +23,8 @@
     package_data=dict(cuenca=['py.typed']),
     python_requires='>=3.6',
     install_requires=[
-        'requests>=2.24,<2.28',
-        'cuenca-validations>=0.9.0,<0.10.0',
+        'requests>=2.24,<2.30',
```
Comment (matin) [inline setup.py:26]: why `<2.30`? The latest version of `requests` is `2.27.1`

**Review by pachCode — APPROVED** [review]

**Pull request merged**